edgelet/edgelet-http/src/lib.rs

```diff
@@ -69,6 +77,103 @@ const TCP_SCHEME: &str = "tcp";
+    key: Option<Vec<u8>>,
+    username: Option<String>,
+    password: Option<String>,
+}
+
+impl PemCertificate {
+    pub fn new(
+        cert: Vec<u8>,
+        key: Option<Vec<u8>>,
+        username: Option<String>,
+        password: Option<String>,
+    ) -> Self {
+        PemCertificate {
+            cert,
+            key,
+            username,
+            password,
+        }
+    }
+
+    pub fn get_certificate(&self) -> &[u8] {
+        &self.cert
+    }
+
+    pub fn from<C: Certificate>(id_cert: &C) -> Result<Self, Error> {
+        let cert = id_cert
+            .pem()
+            .map(|cert_buffer| cert_buffer.as_ref().to_owned())
+            .context(ErrorKind::IdentityCertificate)?;
+
+        let key = match id_cert.get_private_key() {
+            Ok(Some(PrivateKey::Ref(ref_))) => Some(ref_.into_bytes().clone()),
+            Ok(Some(PrivateKey::Key(KeyBytes::Pem(buffer)))) => Some(buffer.as_ref().to_vec()),
+            _ => None,
+        };
+
+        Ok(PemCertificate::new(cert, key, None, None))
+    }
+
+    pub fn get_identity(&self) -> Result<Identity, Error> {
+        let mut certs =
+            X509::stack_from_pem(&self.cert).with_context(|_| ErrorKind::IdentityCertificate)?;
+
+        // the first cert is the identity cert and the other certs are part of the CA
+        // chain; we skip the server cert and build an OpenSSL cert stack with the
+        // other certs
+        let mut ca_certs = Stack::new().with_context(|_| ErrorKind::IdentityCertificate)?;
+        for cert in certs.split_off(1) {
+            ca_certs
+                .push(cert)
+                .with_context(|_| ErrorKind::IdentityCertificate)?;
+        }
+
+        let key = match &self.key {
+            Some(k) => PKey::private_key_from_pem(&k),
+            None => return Err(Error::from(ErrorKind::IdentityPrivateKey)),
+        }?;
+
+        let identity_cert = &certs[0];
+
+        let mut builder = Pkcs12::builder();
+        builder.ca(ca_certs);
+        let pkcs_certs = builder
+            .build(
+                self.password.as_ref().map_or("", String::as_str),
+                self.username.as_ref().map_or("", String::as_str),
+                &key,
+                &identity_cert,
+            )
+            .with_context(|_| ErrorKind::IdentityCertificate)?;
+
+        let der = pkcs_certs
+            .to_der()
+            .with_context(|_| ErrorKind::IdentityCertificate)?;
+
+        let identity = Identity::from_pkcs12(&der, "")?;
+        Ok(identity)
+    }
+}
+
+impl Debug for PemCertificate {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // do not print either the username, password or private key
+        write!(f, "Certificate: {:?}", self.cert)
+    }
+}
+
+impl Display for PemCertificate {
+    // do not print either the username, password or private key
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "Certificate: {:?}", self.cert)
```

Comment:
Is it intentional that you use `Vec<u8>`'s debug formatting here? It doesn't seem to be in the spirit of `Display`, ie to provide a human-readable string.

Reply:
Your comment is valid, to make this more human readable, I guess a way to do this was to parse some attributes of the certificate and print it. Would that work?

Reply:
Is a `Display` impl needed at all?

Reply:
No not really, I can remove this.